packages/block-editor/src/components/block-drop-zone/index.js

```diff
@@ -40,49 +30,67 @@ const parseDropEvent = ( event ) => {
 	return result;
 };
 
-class BlockDropZone extends Component {
-	constructor() {
-		super( ...arguments );
-
-		this.onFilesDrop = this.onFilesDrop.bind( this );
-		this.onHTMLDrop = this.onHTMLDrop.bind( this );
-		this.onDrop = this.onDrop.bind( this );
+export default function useBlockDropZone( { element, clientId, rootClientId } ) {
+	function selector( select ) {
+		const {
+			getBlockIndex,
+			getClientIdsOfDescendants,
+			getSettings,
+			getTemplateLock,
+		} = select( 'core/block-editor' );
+		return {
+			getBlockIndex,
+			getClientIdsOfDescendants,
+			hasUploadPermissions: !! getSettings().mediaUpload,
+			isLockedAll: getTemplateLock( rootClientId ) === 'all',
+		};
 	}
 
-	getInsertIndex( position ) {
-		const { clientId, rootClientId, getBlockIndex } = this.props;
+	const {
+		getBlockIndex,
+		getClientIdsOfDescendants,
+		hasUploadPermissions,
+		isLockedAll,
+	} = useSelect( selector, [ rootClientId ] );
```

Comment:
What pattern did we settle on? I thought we decided to use inline functions?

Reply:
I thought we settled on allowing it. It's especially handy if variable names are clashing. It's also an optimisation if `useSelect` has no dependencies, you can move it out of the component. Here we depend on `rootClientId` though.